python/src/etos_api/routers/testrun/router.py

```diff
@@ -0,0 +1,303 @@
+    :rtype: list
+    """
+    try:
+        suite = requests.get(test_suite_url, timeout=60)
+        suite.raise_for_status()
+    except Exception as exception:  # pylint:disable=broad-except
+        raise AssertionError(f"Unable to download suite from {test_suite_url}") from exception
+    return suite.json()
+
+
+async def validate_suite(test_suite: list[dict[str, Any]]) -> None:
+    """Validate the ETOS test suite through the SuiteValidator.
+
+    :param test_suite_url: The URL to the test suite to validate.
+    """
+    span = trace.get_current_span()
+
+    try:
+        await SuiteValidator().validate(test_suite)
+    except AssertionError as exception:
+        LOGGER.error("Test suite validation failed!")
+        LOGGER.error(exception)
+        span.add_event("Test suite validation failed")
+        raise HTTPException(
+            status_code=400, detail=f"Test suite validation failed. {exception}"
+        ) from exception
+
+
+async def _create_testrun(etos: StartTestrunRequest, span: Span) -> dict:
+    """Create a testrun for ETOS to execute.
+
+    :param etos: Testrun pydantic model.
+    :param span: An opentelemetry span for tracing.
+    :return: JSON dictionary with response.
+    """
+    testrun_id = str(uuid4())
+    LOGGER.identifier.set(testrun_id)
+    span.set_attribute("etos.id", testrun_id)
+
+    LOGGER.info("Download test suite.")
+    test_suite = await download_suite(etos.test_suite_url)
+    LOGGER.info("Test suite downloaded.")
+
+    LOGGER.info("Validating test suite.")
+    await validate_suite(test_suite)
+    LOGGER.info("Test suite validated.")
+
+    etos_library = ETOS("ETOS API", os.getenv("HOSTNAME", "localhost"), "ETOS API")
+
+    LOGGER.info("Get artifact created %r", (etos.artifact_identity or str(etos.artifact_id)))
+    try:
+        artifact = await wait_for_artifact_created(
+            etos_library, etos.artifact_identity, etos.artifact_id
+        )
+    except Exception as exception:  # pylint:disable=broad-except
+        LOGGER.critical(exception)
+        raise HTTPException(
+            status_code=400, detail=f"Could not connect to GraphQL. {exception}"
+        ) from exception
+    if artifact is None:
+        identity = etos.artifact_identity or str(etos.artifact_id)
+        raise HTTPException(
+            status_code=400,
+            detail=f"Unable to find artifact with identity '{identity}'",
+        )
+    LOGGER.info("Found artifact created %r", artifact)
+    # There are assumptions here. Since "edges" list is already tested
+    # and we know that the return from GraphQL must be 'node'.'meta'.'id'
+    # if there are "edges", this is fine.
+    # Same goes for 'data'.'identity'.
+    artifact_id = artifact[0]["node"]["meta"]["id"]
+    identity = artifact[0]["node"]["data"]["identity"]
+    span.set_attribute("etos.artifact.id", artifact_id)
+    span.set_attribute("etos.artifact.identity", identity)
+
+    try:
+        name = test_suite[0].get("name")
```

Comment:
Is it safe to assume the first element in the list is the one we're interested in? Even If so, we should still add code comment.

Reply:
Not really no, but until we have another input I believe we should try to use at least one name.
Added a comment.